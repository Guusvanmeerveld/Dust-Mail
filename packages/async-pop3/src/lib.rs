mod constants;
mod parse;
mod socket;
pub mod types;
mod utils;

use async_native_tls::{TlsConnector, TlsStream};
use parse::{parse_capabilities, Parser};
use socket::Socket;

use tokio::{
    io::{AsyncRead, AsyncWrite},
    net::{TcpStream, ToSocketAddrs},
    time::{timeout, Duration},
};
use types::{
    Capabilities, Capability, Error, ErrorKind, Result, Stats, StatsResponse, UniqueIDResponse,
};

use utils::create_command;

#[derive(Eq, PartialEq, Debug)]
pub enum ClientState {
    Authentication,
    Transaction,
    Update,
    None,
}

pub struct Client<S: AsyncWrite + AsyncRead + Unpin> {
    socket: Option<Socket<S>>,
    capabilities: Capabilities,
    marked_as_del: Vec<u32>,
    greeting: Option<String>,
    read_greeting: bool,
    pub state: ClientState,
}

fn get_connection_timeout(timeout: Option<Duration>) -> Duration {
    match timeout {
        Some(timeout) => timeout,
        None => Duration::from_secs(60),
    }
}

/// Creates a client from a given socket connection.
async fn create_client_from_socket<S: AsyncRead + AsyncWrite + Unpin>(
    socket: Socket<S>,
) -> Result<Client<S>> {
    let mut client = Client {
        marked_as_del: Vec::new(),
        capabilities: Vec::new(),
        greeting: None,
        read_greeting: false,
        socket: Some(socket),
        state: ClientState::Authentication,
    };

    client.greeting = Some(client.read_greeting().await?);

    client.capabilities = client.capa().await?;

    Ok(client)
}

/// Creates a new pop3 client from an existing stream.
/// # Examples
/// ```rust
/// extern crate pop3;
/// use std::net::TcpStream;
///
/// fn main() {
///     // Not recommended to use plaintext, just an example.
///     let stream = TcpStream::connect(("outlook.office365.com", 110)).unwrap();
///
///     let mut client = pop3::new(stream).unwrap();
///
///     client.quit().unwrap();
/// }
/// ```
pub async fn new<S: AsyncRead + AsyncWrite + Unpin>(
    stream: S,
    timeout: Option<Duration>,
) -> Result<Client<S>> {
    let socket = Socket::new(stream, timeout);

    create_client_from_socket(socket).await
}

/// Create a new pop3 client with a tls connection.
pub async fn connect<A: ToSocketAddrs>(
    addr: A,
    domain: &str,
    tls_connector: &TlsConnector,
    connection_timeout: Option<Duration>,
) -> Result<Client<TlsStream<TcpStream>>> {
    let connection_timeout = get_connection_timeout(connection_timeout);

    let tcp_stream = timeout(connection_timeout, TcpStream::connect(addr)).await??;

    let tls_stream = tls_connector.connect(domain, tcp_stream).await?;

    let socket = Socket::new(tls_stream, None);

    create_client_from_socket(socket).await
}

/// Creates a new pop3 client using a plain connection.
///
/// DO NOT USE in a production environment. Your password will be sent over a plain tcp stream which hackers could intercept.
pub async fn connect_plain<A: ToSocketAddrs>(
    addr: A,
    connection_timeout: Option<Duration>,
) -> Result<Client<TcpStream>> {
    let connection_timeout = get_connection_timeout(connection_timeout);

    let tcp_stream = timeout(connection_timeout, TcpStream::connect(addr)).await??;

    let socket = Socket::new(tcp_stream, None);

    create_client_from_socket(socket).await
}

impl<S: AsyncRead + AsyncWrite + Unpin> Client<S> {
    /// Check if the client is in the correct state and return a mutable reference to the tcp connection.
    fn get_socket_mut(&mut self) -> Result<&mut Socket<S>> {
        match self.socket.as_mut() {
            Some(socket) => {
                if self.state == ClientState::Transaction
                    || self.state == ClientState::Authentication
                {
                    Ok(socket)
                } else {
                    Err(Error::new(
                        ErrorKind::ShouldNotBeConnected,
                        "There is a connection, but our state indicates that we should not be connected",
                    ))
                }
            }
            None => Err(Error::new(
                ErrorKind::NotConnected,
                "Not connected to any server",
            )),
        }
    }

    /// Check if the client is in the correct state.
    fn check_client_state(&self, state: ClientState) -> Result<()> {
        if self.state != state {
            Err(Error::new(
                ErrorKind::IncorrectStateForCommand,
                "The connection is not the right state to use this command",
            ))
        } else {
            Ok(())
        }
    }

    /// ## Current client state
    ///
    /// Indicates what state the client is currently in, can be either
    /// Authentication, Transaction, Update or None.
    ///
    /// Some methods are only available in some specified states and will error if run in an incorrect state.
    ///
    /// https://www.rfc-editor.org/rfc/rfc1939#section-3
    pub fn get_state(&self) -> &ClientState {
        &self.state
    }

    /// ## NOOP
    /// The POP3 server does nothing, it merely replies with a positive response.
    /// ### Arguments: none
    /// ### Restrictions:
    /// - May only be given in the TRANSACTION state
    /// ### Possible Responses:
    /// - OK
    /// # Examples:
    /// ```rust,ignore
    /// client.noop()?;
    /// ```
    /// https://www.rfc-editor.org/rfc/rfc1939#page-9
    pub async fn noop(&mut self) -> Result<()> {
        let socket = self.get_socket_mut()?;

        let command = "NOOP";

        socket.send_command(command, false).await?;

        Ok(())
    }

    pub async fn uidl(&mut self, msg_number: Option<u32>) -> Result<UniqueIDResponse> {
        self.check_capability(vec![Capability::Uidl])?;

        match msg_number.as_ref() {
            Some(msg_number) => self.check_deleted(msg_number)?,
            None => {}
        };

        let response_is_multi_line = msg_number.is_none();

        let socket = self.get_socket_mut()?;

        let arguments = if msg_number.is_some() {
            vec![msg_number.unwrap().to_string()]
        } else {
            Vec::new()
        };

        let command = create_command("UIDL", &arguments)?;

        let response = socket.send_command(command, response_is_multi_line).await?;

        let parser = Parser::new(response);

        if response_is_multi_line {
            Ok(UniqueIDResponse::UniqueIDList(parser.to_unique_id_list()))
        } else {
            Ok(UniqueIDResponse::UniqueID(parser.to_unique_id()))
        }
    }

    pub async fn top(&mut self, msg_number: u32, lines: u32) -> Result<Vec<u8>> {
        self.check_deleted(&msg_number)?;

        self.check_capability(vec![Capability::Top])?;

        let socket = self.get_socket_mut()?;

        let command = format!("TOP {} {}", msg_number, lines);

        socket.send_command(command, false).await?;

        let mut response: Vec<u8> = Vec::new();

        socket.read_multi_line(&mut response).await?;

        Ok(response)
    }

    /// Check whether a given message is marked as deleted by the server.
    ///
    /// If this function returns true then the message may still not exist.
    /// # Examples:
    /// ```rust,ignore
    /// let msg_number: u32 = 8;
    /// let is_deleted = client.is_deleted(msg_number);
    /// assert_eq!(is_deleted, false);
    /// ```
    pub fn is_deleted(&mut self, msg_number: &u32) -> bool {
        self.marked_as_del.sort();

        match self.marked_as_del.binary_search(msg_number) {
            Ok(_) => true,
            Err(_) => false,
        }
    }

    fn check_deleted(&mut self, msg_number: &u32) -> Result<()> {
        if self.is_deleted(msg_number) {
            Err(types::Error::new(
                types::ErrorKind::MessageIsDeleted,
                "This message has been marked as deleted and cannot be refenced anymore",
            ))
        } else {
            Ok(())
        }
    }

    /// ## DELE
    /// The POP3 server marks the message as deleted.  Any future reference to the message-number associated with the message in a POP3 command generates an error.  The POP3 server does not actually delete the message until the POP3 session enters the UPDATE state.
    /// ### Arguments:
    /// - a message-number (required) which may NOT refer to a message marked as deleted.
    /// ### Restrictions:
    /// - may only be given in the TRANSACTION state
    /// ### Possible Responses:
    /// - OK: message deleted
    /// - ERR: no such message
    /// # Examples
    /// ```rust,ignore
    /// let msg_number: u32 = 8;
    /// let is_deleted = client.is_deleted(msg_number);
    ///
    /// println!("{}", is_deleted);
    /// ```
    pub async fn dele(&mut self, msg_number: u32) -> Result<()> {
        self.check_deleted(&msg_number)?;

        let socket = self.get_socket_mut()?;

        let command = format!("DELE {}", msg_number);

        socket.send_command(command.as_bytes(), false).await?;

        Ok(())
    }

    /// ## RSET
    /// If any messages have been marked as deleted by the POP3
    /// server, they are unmarked.
    /// ### Arguments: none
    /// ### Restrictions:
    /// - May only be given in the TRANSACTION state
    /// ### Possible Responses:
    /// - +OK
    /// # Examples:
    /// ```rust,ignore
    /// client.rset().unwrap();
    /// ```
    /// https://www.rfc-editor.org/rfc/rfc1939#page-9
    pub async fn rset(&mut self) -> Result<()> {
        let socket = self.get_socket_mut()?;

        let command = b"RSET";

        socket.send_command(command, false).await?;

        Ok(())
    }

    /// ## RETR
    /// Retrieves the full RFC822 compliant message from the server and returns it as a byte vector
    /// ### Arguments:
    /// - A message-number (required) which may NOT refer to a message marked as deleted
    /// ### Restrictions:
    /// - May only be given in the TRANSACTION state
    /// ### Possible Responses:
    /// - OK: message follows
    /// - ERR: no such message
    /// # Examples
    /// ```rust,ignore
    /// extern crate mailparse;
    /// use mailparse::parse_mail;
    ///
    /// let response = client.retr(1).unwrap();
    ///
    /// let parsed = parse_mail(&response);
    ///
    /// let subject = parsed.headers.get_first_value("Subject").unwrap();
    ///
    /// println!("{}", subject);
    /// ```
    /// https://www.rfc-editor.org/rfc/rfc1939#page-8
    pub async fn retr(&mut self, msg_number: u32) -> Result<Vec<u8>> {
        self.check_deleted(&msg_number)?;

        let socket = self.get_socket_mut()?;

        let arguments = vec![msg_number.to_string()];

        let command = create_command("RETR", &arguments)?;

        socket.send_bytes(command.as_bytes()).await?;

        let mut response: Vec<u8> = Vec::new();

        socket.read_multi_line(&mut response).await?;

        Ok(response)
    }

    pub async fn list(&mut self, msg_number: Option<u32>) -> Result<StatsResponse> {
        match msg_number.as_ref() {
            Some(msg_number) => {
                self.check_deleted(msg_number)?;
            }
            None => {}
        };

        let socket = self.get_socket_mut()?;

        let response_is_multi_line = msg_number.is_none();

        let arguments = if !response_is_multi_line {
            vec![msg_number.unwrap().to_string()]
        } else {
            Vec::new()
        };

        let command = create_command("LIST", &arguments)?;

        let response = socket.send_command(command, response_is_multi_line).await?;

        let parser = Parser::new(response);

        if response_is_multi_line {
            // println!("{}", response);

            Ok(StatsResponse::StatsList(parser.to_stats_list()))
        } else {
            Ok(StatsResponse::Stats(parser.to_stats()))
        }
    }

    pub async fn stat(&mut self) -> Result<Stats> {
        let socket = self.get_socket_mut()?;

        let command = b"STAT";

        let response = socket.send_command(command, false).await?;

        let parser = Parser::new(response);

        let stats = parser.to_stats();

        Ok(stats)
    }

    pub async fn apop(&mut self, name: &str, digest: &str) -> Result<()> {
        self.check_client_state(ClientState::Authentication)?;

        self.has_read_greeting()?;

        let socket = self.get_socket_mut()?;

        let command = format!("APOP {} {}", name, digest);

        socket.send_command(command, false).await?;

        self.state = ClientState::Transaction;

        Ok(())
    }

    pub async fn auth<U: AsRef<str>>(&mut self, token: U) -> Result<()> {
        self.check_client_state(ClientState::Authentication)?;

        self.check_capability(vec![Capability::Sasl(vec![String::from("XOAUTH2")])])?;

        self.has_read_greeting()?;

        let socket = self.get_socket_mut()?;

        let command = format!("AUTH {}", token.as_ref());

        socket.send_command(command, false).await?;

        self.state = ClientState::Transaction;

        Ok(())
    }

    pub async fn login<U: AsRef<str>, P: AsRef<str>>(
        &mut self,
        user: U,
        password: P,
    ) -> Result<()> {
        self.check_client_state(ClientState::Authentication)?;

        self.check_capability(vec![
            Capability::User,
            Capability::Sasl(vec![String::from("PLAIN")]),
        ])?;

        self.has_read_greeting()?;

        let socket = self.get_socket_mut()?;

        let command = format!("USER {}", user.as_ref());

        socket.send_command(command, false).await?;

        let command = format!("PASS {}", password.as_ref());

        socket.send_command(command, false).await?;

        self.capabilities = self.capa().await?;

        self.state = ClientState::Transaction;

        Ok(())
    }

    pub async fn quit(&mut self) -> Result<()> {
        let socket = self.get_socket_mut()?;

        let command = b"QUIT";

        socket.send_command(command, false).await?;

        self.state = ClientState::Update;
        self.socket = None;
        self.state = ClientState::None;

        self.marked_as_del.clear();
        self.capabilities.clear();

        Ok(())
    }

    /// Check whether the server supports one of the given capabilities.
    pub fn has_capability(&mut self, capabilities: Vec<Capability>) -> bool {
        self.capabilities.sort();

        match capabilities.iter().find(|capability| {
            match self.capabilities.binary_search(&capability) {
                Ok(_) => true,
                Err(_) => false,
            }
        }) {
            Some(_) => true,
            None => false,
        }
    }

    /// Make sure the given capabilities are present
    fn check_capability(&mut self, capability: Vec<Capability>) -> Result<()> {
        if !self.has_capability(capability) {
            Err(types::Error::new(
                types::ErrorKind::FeatureUnsupported,
                "The remote pop server does not support this command/function",
            ))
        } else {
            Ok(())
        }
    }

    /// Returns the current list of capabilities given by the server.
    pub fn capabilities(&self) -> &Capabilities {
        &self.capabilities
    }

    /// Fetches a list of capabilities for the currently connected server and returns it.
    pub async fn capa(&mut self) -> Result<Capabilities> {
        let socket = self.get_socket_mut()?;

        let command = b"CAPA";

        let response = socket.send_command(command, true).await?;

        Ok(parse_capabilities(&response))
    }

    fn has_read_greeting(&self) -> Result<()> {
        if !self.read_greeting {
            Err(types::Error::new(
                types::ErrorKind::ServerFailedToGreet,
                "Did not connect to the server correctly, as we did not get a greeting yet",
            ))
        } else {
            Ok(())
        }
    }

    async fn read_greeting(&mut self) -> Result<String> {
        assert!(!self.read_greeting, "Cannot read greeting twice");

        let socket = self.get_socket_mut()?;

        let response = socket.read_response(false).await?;

        self.read_greeting = true;
        Ok(response)
    }

    pub fn greeting(&self) -> Option<&str> {
        match &self.greeting {
            Some(greeting) => Some(greeting.as_str()),
            None => None,
        }
    }
}

#[cfg(test)]
mod test;
