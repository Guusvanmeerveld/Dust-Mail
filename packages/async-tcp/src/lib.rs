use async_native_tls::{TlsConnector, TlsStream};
pub use config::Config;
use constants::{CR, LF};
use tokio::{
    io::{AsyncBufReadExt, AsyncRead, AsyncWrite, AsyncWriteExt, BufStream},
    net::{TcpStream, ToSocketAddrs},
};

mod config;
mod constants;
pub mod types;

use types::{Error, ErrorKind, Result};

pub struct TcpClient<S: AsyncWrite + AsyncRead + Unpin> {
    stream: BufStream<S>,
    config: Config,
}

pub fn new<T: AsyncWrite + AsyncRead + Unpin>(stream: T, config: Config) -> TcpClient<T> {
    TcpClient {
        stream: BufStream::new(stream),
        config,
    }
}

/// Connects to a remote server in a secure manner using a given address and hostname, with a timeout duration.
pub async fn connect<A: ToSocketAddrs, H: AsRef<str>>(
    addr: A,
    host: H,
    config: Option<Config>,
) -> Result<TcpClient<TlsStream<TcpStream>>> {
    let config = config.unwrap_or_default();

    let tcp = tokio::time::timeout(config.timeout().clone(), TcpStream::connect(addr)).await??;

    let tls = TlsConnector::new();

    let stream = tls.connect(host.as_ref(), tcp).await?;

    let client = new(stream, config);

    Ok(client)
}

/// Connects to a remote server given address with a timeout duration.
pub async fn connect_plain<A: ToSocketAddrs>(
    addr: A,
    config: Option<Config>,
) -> Result<TcpClient<TcpStream>> {
    let config = config.unwrap_or_default();

    let stream = tokio::time::timeout(config.timeout().clone(), TcpStream::connect(addr)).await??;

    let client = new(stream, config);

    Ok(client)
}

impl<S: AsyncWrite + AsyncRead + Unpin> TcpClient<S> {
    /// Creates a new client using a given stream and timeout duration.

    /// Expects a single line response
    pub async fn send_command<C: AsRef<[u8]>>(&mut self, command: C) -> Result<String> {
        self.send_bytes(command).await?;

        let response = self.read_response().await?;

        Ok(response)
    }

    /// Write some bytes to the socket, end it off with a LF CR and send it to the remote server.
    pub async fn send_bytes<B: AsRef<[u8]>>(&mut self, bytes: B) -> Result<()> {
        self.stream.write_all(bytes.as_ref()).await?;
        self.stream.write_all(&[LF, CR]).await?;
        self.stream.flush().await?;

        Ok(())
    }

    /// Reads a single line response
    pub async fn read_response(&mut self) -> Result<String> {
        let mut response = String::new();

        let bytes_read = tokio::time::timeout(
            self.config.timeout().clone(),
            self.stream.read_line(&mut response),
        )
        .await??;

        if bytes_read < 1 {
            return Err(Error::new(
                ErrorKind::EmptyResponse,
                "Server did not send any bytes",
            ));
        }

        Ok(response)
    }

    pub async fn close(&mut self) -> Result<()> {
        self.stream.shutdown().await?;

        Ok(())
    }
}
