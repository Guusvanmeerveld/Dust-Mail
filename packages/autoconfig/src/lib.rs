use dust_mail_utils::validate_email;

use http::Client;
use types::{config::Config, Result};

mod http;
mod parse;
pub mod types;
mod utils;

const AT_SYMBOL: char = '@';

/// Given an email providers domain, try to connect to autoconfig servers for that provider and return the config.
pub async fn from_domain<D: AsRef<str>>(domain: D) -> Result<Option<Config>> {
    let client = Client::new()?;

    let urls = vec![
        // Try connect to connect with the users mail server directly
        format!("http://autoconfig.{}/mail/config-v1.1.xml", domain.as_ref()),
        // The fallback url
        format!(
            "http://{}/.well-known/autoconfig/mail/config-v1.1.xml",
            domain.as_ref()
        ),
        // If the previous two methods did not work then the email server provider has not setup Thunderbird autoconfig, so we ask Mozilla for their config.
        format!(
            "https://autoconfig.thunderbird.net/v1.1/{}",
            domain.as_ref()
        ),
    ];

    let config_unparsed: Option<String> = client.request_urls(urls).await;

    match config_unparsed {
        Some(config_unparsed) => {
            let config = parse::from_str(&config_unparsed)?;

            Ok(Some(config))
        }
        None => Ok(None),
    }
}

/// Given an email address, try to connect to the email providers autoconfig servers and return the config that was found.
pub async fn from_addr(email_address: &str) -> Result<Option<Config>> {
    if !validate_email(email_address) {
        return Err(types::Error::new(
            types::ErrorKind::BadInput,
            "Given email address is invalid",
        ));
    };

    let mut split = email_address.split(AT_SYMBOL);

    // Skip the prefix
    split.next();

    let domain = match split.next() {
        Some(domain) => domain,
        None => {
            return Err(types::Error::new(
                types::ErrorKind::BadInput,
                "An email address must specify a domain after the '@' symbol",
            ))
        }
    };

    from_domain(domain).await
}

#[cfg(test)]
mod test;
