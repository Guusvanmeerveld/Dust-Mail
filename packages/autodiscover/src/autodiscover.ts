import axios from "axios";

import dns from "dns/promises";

import AutodiscoverResponse from "./interfaces/emailServer";

interface Credentials {
	username: string;
	password: string;
}

const fetchServerWithAutodiscover = async (
	server: string,
	credentials?: Credentials
): Promise<AutodiscoverResponse | undefined> => {
	const srv: {
		name: string;
		port: number;
		priority: number;
		weight: number;
	}[] = await dns.resolveSrv(`_autodiscover._tcp.${server}`).catch(() => []);

	let autoDiscoverDomain: string;

	if (srv.length != 0) {
		const response = srv.reduce((max, item) =>
			item.priority > max.priority ? item : max
		);

		await fetchDataFromAutodiscoverURL(
			response.name,
			response.port,
			credentials
		);

		return;
	} else {
		await fetchDataFromAutodiscoverURL(server, 443, credentials);
		return;
	}
};

const fetchDataFromAutodiscoverURL = async (
	server: string,
	port: number,
	credentials?: Credentials
) => {
	const data = await axios(
		`${
			port == 443 ? "https" : "http"
		}://${server}/Autodiscover/Autodiscover.svc`,
		{
			method: "POST",
			headers: { "Content-Type": "text/xml; charset=utf-8" }
		}
	).catch((error) => {
		// console.log(error);
	});

	// console.log(data);
};

export default fetchServerWithAutodiscover;
