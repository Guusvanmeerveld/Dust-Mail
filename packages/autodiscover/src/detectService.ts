import net from "net";

import tls from "tls";

import { IncomingServer, OutgoingServer } from "./interfaces/emailServer";

import ServerConfig from "./interfaces/serverConfig";

/**
 * Will take a config to connect with an email related server and detect what kind of email service is running on that server
 */
export const detectServiceFromConfig = async (
	config: ServerConfig
): Promise<IncomingServer | OutgoingServer> => {
	const responseFromImapCommand: string = await connectToServerAndSendCommand(
		config,
		"abcd CAPABILITY"
	);

	if (responseFromImapCommand) {
		const service = detectServiceFromResponse(responseFromImapCommand);

		if (service) return service;
	}

	throw new Error("Not found");
};

const detectServiceFromResponse = (
	response: string
): IncomingServer | OutgoingServer | void => {
	const split = response.toLowerCase().split(" ");

	const isImap = split.find((string) => string.match(/imap/));

	if (isImap) return "imap";

	const isSmtp = split.find((string) => string.match(/smtp/));

	if (isSmtp) return "smtp";

	const isPop = split.find((string) => string.match(/pop/));

	if (isPop) return "pop3";
};

const connectToServerAndSendCommand = (
	config: ServerConfig,
	command?: string
): Promise<string> =>
	new Promise((resolve, reject) => {
		const connection =
			config.security == "TLS"
				? tls.connect({
						port: config.port,
						host: config.server
				  })
				: net.connect({
						port: config.port,
						host: config.server
				  });

		connection.on("close", reject);

		connection.on("error", reject);

		connection.on("connect", () => {
			if (command)
				connection.write(command, () =>
					setTimeout(() => {
						connection.destroy();
						reject();
					}, 10000)
				);
		});

		connection.on("data", (chunk) => {
			connection.destroy();
			resolve(chunk.toString());
		});
	});
