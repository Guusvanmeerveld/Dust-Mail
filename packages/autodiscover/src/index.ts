import fetchServerWithAutodiscover from "autodiscover";
import validator from "validator";

import AutodiscoverResponse from "./interfaces/emailServer";

const fetchServerFromEmail = async (
	email: string,
	password?: string
): Promise<AutodiscoverResponse> => {
	const isValidEmail = validator.isEmail(email);

	if (!isValidEmail) throw new Error("Email adress is not valid");

	const server = email.split("@").pop() as string;

	let response: AutodiscoverResponse | undefined;

	response = await fetchServerWithAutodiscover(
		server,
		password ? { username: email, password } : undefined
	);

	if (!response)
		throw new Error(`Could not find any email servers related to ${email}`);

	return response;
};

export default fetchServerFromEmail;
