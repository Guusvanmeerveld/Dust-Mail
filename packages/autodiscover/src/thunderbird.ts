import { XMLParser } from "fast-xml-parser";

import { OutgoingServiceType, IncomingServiceType } from "@dust-mail/typings";

import AutoConfigFile, { Server } from "./interfaces/autoConfigFile";
import AutodiscoverResponse, {
	EmailServer,
	Security
} from "./interfaces/emailServer";

import axios from "axios";

const fetchServerWithThunderbird = async (
	server: string,
	email: string
): Promise<AutodiscoverResponse | undefined> => {
	const emailProviderUrl = `http://autoconfig.${server}/mail/config-v1.1.xml?emailaddress=${email}`;

	const emailProviderResponse = await axios
		.get<string>(emailProviderUrl, {
			method: "GET"
		})
		.catch(() => {
			return;
		});

	if (emailProviderResponse && emailProviderResponse.data) {
		const data = parseAutoConfigFile(emailProviderResponse.data);

		if (data) return data;
	}

	const emailProviderBackupUrl = `http://${server}/.well-known/autoconfig/mail/config-v1.1.xml`;

	const emailProviderBackupResponse = await axios
		.get<string>(emailProviderBackupUrl, {
			method: "GET"
		})
		.catch(() => {
			return;
		});

	if (emailProviderBackupResponse && emailProviderBackupResponse.data) {
		const data = parseAutoConfigFile(emailProviderBackupResponse.data);

		if (data) return data;
	}

	const mozillaServerUrl = `https://autoconfig.thunderbird.net/v1.1/${server}`;

	const mozillaServerResponse = await axios
		.get<string>(mozillaServerUrl, {
			method: "GET"
		})
		.catch(() => {
			return;
		});

	if (mozillaServerResponse && mozillaServerResponse.data) {
		const data = parseAutoConfigFile(mozillaServerResponse.data);

		if (data) return data;
	}

	return;
};

const parseAutoConfigFile = (
	autoConfigFile: string
): AutodiscoverResponse | undefined => {
	const parser = new XMLParser({
		ignoreAttributes: false,
		attributeNamePrefix: "",
		attributesGroupName: "@"
	});

	const data: AutoConfigFile | undefined = parser.parse(autoConfigFile);

	if (!data?.clientConfig?.emailProvider) return;

	const incomingServers = data.clientConfig.emailProvider.incomingServer;

	let incomingServer: Server | undefined;

	if (Array.isArray(incomingServers)) {
		if (incomingServers.length != 0) incomingServer = incomingServers[0];
	} else incomingServer = incomingServers;

	const incomingServerType = incomingServer?.["@"].type as IncomingServiceType;

	const outgoingServers = data.clientConfig.emailProvider.outgoingServer;

	let outgoingServer: Server | undefined;

	if (Array.isArray(outgoingServers)) {
		if (outgoingServers.length != 0) outgoingServer = outgoingServers[0];
	} else outgoingServer = outgoingServers;

	const outgoingServerType = outgoingServer?.["@"].type as OutgoingServiceType;

	return [
		createEmailServer(incomingServerType, incomingServer),
		createEmailServer(outgoingServerType, outgoingServer)
	];
};

function createEmailServer(
	type: IncomingServiceType,
	server?: Server
): EmailServer<IncomingServiceType>;
function createEmailServer(
	type: OutgoingServiceType,
	server?: Server
): EmailServer<OutgoingServiceType>;
function createEmailServer(
	type: IncomingServiceType | OutgoingServiceType,
	server?: Server
): EmailServer<IncomingServiceType | OutgoingServiceType> | undefined {
	if (!server) return;

	let security: Security;

	switch (server.socketType) {
		case "SSL":
			security = "TLS";
			break;

		case "STARTTLS":
			security = "STARTTLS";
			break;

		case "plain":
			security = "NONE";
			break;
	}

	return {
		port: server.port,
		server: server.hostname,
		security,
		type
	};
}

export default fetchServerWithThunderbird;
