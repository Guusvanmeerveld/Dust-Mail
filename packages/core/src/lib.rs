mod client;
mod detect;
mod tls;
mod utils;

#[cfg(feature = "imap")]
mod imap;

#[cfg(feature = "pop")]
mod pop;

pub mod types;

pub use client::{ClientConstructor, ClientType, LoginOptions};
pub use detect::ServiceDetector;
