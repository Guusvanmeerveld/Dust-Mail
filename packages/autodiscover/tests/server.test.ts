import { expect, test } from "vitest";

import fetchServerWithThunderbird from "../src/thunderbird";

import { detectServiceFromConfig } from "../src/detectService";

// import fetchServerWithAutodiscover from "../autodiscover";

const username = "mail@guusvanmeerveld.dev";
const server = username.split("@").pop() as string;

// test("Test autodiscover functionality", async () => {
// 	await fetchServerWithAutodiscover(server)
// 		.catch((error: Error) =>
// 			expect(error.message).toBe("Email adress is not valid")
// 		)
// 		.then((data) => expect(data).toBeUndefined());
// });

test("Test Thunderbird autoconfig discovery functionality", async () => {
	await fetchServerWithThunderbird(server, username).then((data) => {
		expect(data).toMatchSnapshot();
	});
});

test("Detect imap service from config", async () => {
	await detectServiceFromConfig({
		port: 993,
		server: "imap.guusvanmeerveld.dev",
		security: "TLS"
	}).then((data) => expect(data).toBe("imap"));
});

test("Detect smtp service from config", async () => {
	await detectServiceFromConfig({
		port: 465,
		server: "smtp.guusvanmeerveld.dev",
		security: "TLS"
	}).then((data) => expect(data).toBe("smtp"));
});

test("Detect pop3 service from config", async () => {
	await detectServiceFromConfig({
		port: 995,
		server: "pop3.live.com",
		security: "TLS"
	}).then((data) => expect(data).toBe("pop3"));
});
