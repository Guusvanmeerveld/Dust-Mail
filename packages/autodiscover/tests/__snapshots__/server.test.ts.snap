// Vitest Snapshot v1

exports[`Test Thunderbird autoconfig discovery functionality 1`] = `
[
  {
    "port": 993,
    "security": "TLS",
    "server": "imap.guusvanmeerveld.dev",
    "type": "imap",
  },
  {
    "port": 587,
    "security": "STARTTLS",
    "server": "smtp.guusvanmeerveld.dev",
    "type": "smtp",
  },
]
`;

exports[`Validate autodiscover checker Test Thunderbird autoconfig discovery functionality 1`] = `
Array [
  Object {
    "port": 993,
    "security": "TLS",
    "server": "imap.guusvanmeerveld.dev",
    "type": "imap",
  },
  Object {
    "port": 587,
    "security": "STARTTLS",
    "server": "smtp.guusvanmeerveld.dev",
    "type": "smtp",
  },
]
`;
